themes/src/main/resources/theme/keycloak-preview/account/resources/app/content/signingin-page/SigningInPage.tsx

```diff
@@ -68,116 +92,202 @@ class SigningInPage extends React.Component<SigningInPageProps, SigningInPageSta
     public constructor(props: SigningInPageProps) {
         super(props);
         this.state = {
-            twoFactorEnabled: true,
-            twoFactorEnabledText: Msg.localize('twoFactorEnabled'),
-            isTotpConfigured: features.isTotpConfigured,
+            credentialContainers: new Map()
         }
-        this.setLastPwdUpdate();
+
+        this.getCredentialContainers();
     }
 
-    private setLastPwdUpdate(): void {
-        AccountServiceClient.Instance.doGet("/credentials/password")
-            .then((response: AxiosResponse<PasswordDetails>) => {
-                if (response.data.lastUpdate) {
-                    const lastUpdate: number = response.data.lastUpdate;
-                    this.setState({lastPasswordUpdate: lastUpdate});
-                }
+    private getCredentialContainers(): void {
+        AccountServiceClient.Instance.doGet("/credentials")
+            .then((response: AxiosResponse<CredentialContainer[]>) => {
+
+                const allContainers: CredContainerMap = new Map();
+                response.data.forEach(container => {
+                    let categoryMap = allContainers.get(container.category);
+                    if (!categoryMap) {
+                        categoryMap = new Map();
+                        allContainers.set(container.category, categoryMap);
+                    }
+                    categoryMap.set(container.type, container);
+                });
+
+                this.setState({credentialContainers: allContainers});
             });
     }
     
-    private handleTwoFactorSwitch = () => {
-        if (this.state.twoFactorEnabled) {
-            this.setState({twoFactorEnabled: false, twoFactorEnabledText: Msg.localize('twoFactorDisabled')})
-        } else {
-            this.setState({twoFactorEnabled: true, twoFactorEnabledText:  Msg.localize('twoFactorEnabled')})
-        }
-    }
-
-    private handleRemoveTOTP = () => {
-      AccountServiceClient.Instance.doDelete("/totp/remove")
+    private handleRemove = (credentialId: string, userLabel: string) => {
+      AccountServiceClient.Instance.doDelete("/credentials/" + credentialId)
         .then(() => {
-            this.setState({isTotpConfigured: false});
-            ContentAlert.success('successTotpRemovedMessage');
+            this.getCredentialContainers();
+            ContentAlert.success('successRemovedMessage', [userLabel]);
         });
     }
 
     public render(): React.ReactNode {
```

Comment:
It looks a bit weird, or maybe too technical, that there are two "passwords" on top of each other. Again, IMHO ok for tech preview.
![Screenshot 2020-01-22 at 18 11 47](https://user-images.githubusercontent.com/13906855/72916644-ae528180-3d42-11ea-9b92-2e77eea5a55a.png)

Reply:
I wasn't totally comfortable with that either.  Right now that's just a placeholder so that it's consistent with the other types.  I'm open to suggestions.

Reply:
I suggest to keep it like this and possibly revisit later based on the feedback from users...

For the future, I wonder if we may have 2 categories instead of 3. Instead of "Password, "Two-Factor Authentication" and "Passwordless", we may have just something like "1st factor authentication" and "Two-factor authentication" . From the built-in credentials, the "Password" and "WebAuthn Passwordless" will be 1st-factor Category. And "OTP" and "WebAuthn" will be two-factor category. We may probably wait for the feedback from users and customers whether to do something like this...

Reply:
> I suggest to keep it like this and possibly revisit later based on the feedback from users...

I'm not sure we can expect this kind of feedback ("Hey, it looks weird!"). :) IMHO it is about the "look and feel", about the impression from the admin console and most importantly how easy-to-use and intuitive it is for an end user that doesn't have any technical background. And I don't think those kind of users will file a JIRA. :) But I'm no UX expert, so it's just how I feel it. ;)

> For the future, I wonder if we may have 2 categories instead of 3. Instead of "Password, "Two-Factor Authentication" and "Passwordless", we may have just something like "1st factor authentication" and "Two-factor authentication" . From the built-in credentials, the "Password" and "WebAuthn Passwordless" will be 1st-factor Category. And "OTP" and "WebAuthn" will be two-factor category. We may probably wait for the feedback from users and customers whether to do something like this...

This would even more strongly imply that 2FA is an independent credential type which it isn't.

Reply:
@vmuzikar Please file a JIRA about this and reference this discussion.  I think it's worth taking another look and I wouldn't want us to forget about it after tech preview.

Reply:
Created https://issues.redhat.com/browse/KEYCLOAK-12781